alf/algorithms/algorithm.py

```diff
@@ -217,17 +264,24 @@ def _get_opt_and_var_sets(self):
                 else:
                     raise ValueError("Unsupported module type %s" % module)
             opt_and_var_sets.append((opt, vars))
+        # Check that each variable is optimized by only one optimizer
+        var_ids = [i for _, vars in opt_and_var_sets for i in map(id, vars)]
+        for _, vars in opt_and_var_sets:
+            for v in vars:
+                assert var_ids.count(id(v)) == 1, \
+                    ("Variable '%s' is optimized by multiple optimizers!" % v.name)
         return opt_and_var_sets
 
     def get_optimizer_info(self):
         """Return the optimizer info for all the modules in a string.
         """
-        optimizer_and_module_sets = self.get_optimizer_and_module_sets()
+        optimizer_and_module_sets, _ = self.get_optimizer_and_module_sets()
         optimizer_info = []
-        for opt, module_set in optimizer_and_module_sets:
+        for alg_name, opt, module_set in optimizer_and_module_sets:
```

Comment:
opt may be None  when there is  no optimizer for root algorithm , should filter out `opt == None`

Reply:
The assertion inside `train_complete()` will make sure that each opt is not None: `assert optimizer is not None, "optimizer needs to be provides at __init__()"`

Reply:
changed